### PR TITLE
Generalize boundary condition options

### PR DESCRIPTION
**Pull request by streeve**
1) Separates geometry from region for BC; adds cylinder as additional example
2) Simplification for cases of only one plane
3) Unify interface to use time in all BC
4) Adds options for user-defined regions or a boundary index space (for full customization)

Currently partially embedded in #108

### CODE REVIEW

File: examples/mechanics/crack_branching.cpp [hunk 1]
```diff
@@ -90,17 +90,18 @@ void crackBranchingExample( const std::string filename )
...
-                                     high_corner[1] - dy, high_corner[1] + dy,
-                                     low_corner[2], high_corner[2] );
-    std::vector<CabanaPD::RegionBoundary> planes = { plane1, plane2 };
+    CabanaPD::RegionBoundary<CabanaPD::RectangularPrism> plane1(
+        low_corner[0], high_corner[0], low_corner[1] - dy, low_corner[1] + dy,
+        low_corner[2], high_corner[2] );
+    CabanaPD::RegionBoundary<CabanaPD::RectangularPrism> plane2(
+        low_corner[0], high_corner[0], high_corner[1] - dy, high_corner[1] + dy,
+        low_corner[2], high_corner[2] );
+    std::vector<CabanaPD::RegionBoundary<CabanaPD::RectangularPrism>> planes = {
```
Comment (pabloseleson) [inline examples/mechanics/crack_branching.cpp:99]: What happens if we could like to combine a RectangularPrism region with a different type of region?

File: examples/mechanics/crack_branching.cpp [hunk 2]
```diff
@@ -90,17 +90,18 @@ void crackBranchingExample( const std::string filename )
...
+        plane1, plane2 };
     auto particles_f = particles->getForce();
     auto particles_x = particles->getReferencePosition();
     // Create a symmetric force BC in the y-direction.
-    auto bc_op = KOKKOS_LAMBDA( const int pid )
+    auto bc_op = KOKKOS_LAMBDA( const int pid, const double )
```
Comment (pabloseleson) [inline examples/mechanics/crack_branching.cpp:104]: What is the extra entry "const double" for?

File: src/CabanaPD_Boundary.hpp [hunk 3]
```diff
@@ -40,14 +55,53 @@ struct RegionBoundary
...
+
+// Define a subset of the system as the boundary with a cylinder.
+template <>
+struct RegionBoundary<Cylinder>
+{
+    double radius_in;
+    double radius_out;
+    double low_z;
+    double high_z;
+    double x_center = 0.0;
```
Comment (pabloseleson) [inline src/CabanaPD_Boundary.hpp:77]: Do we have to hardcode x_center and y_center?

File: src/CabanaPD_Boundary.hpp [hunk 4]
```diff
@@ -40,14 +55,53 @@ struct RegionBoundary
...
+        : radius_in( _radius_in )
+        , radius_out( _radius_out )
+        , low_z( _low_z )
+        , high_z( _high_z ){};
+
+    template <class PositionType>
+    KOKKOS_INLINE_FUNCTION bool inside( const PositionType& x,
+                                        const int pid ) const
+    {
+        double rsq = ( x( pid, 0 ) - x_center ) * ( x( pid, 0 ) - x_center ) +
```
Comment (pabloseleson) [inline src/CabanaPD_Boundary.hpp:91]: Would it be possible to pass a (general) user-defined function, instead of defining a set of pre-defining  geometries?

File: src/CabanaPD_Boundary.hpp [hunk 5]
```diff
@@ -304,12 +375,59 @@ auto createBoundaryCondition( UserFunctor user_functor, ExecSpace exec_space,
...
+    using memory_space = typename BoundaryParticles::memory_space;
+    using bc_index_type = BoundaryIndexSpace<memory_space, Custom>;
+    bc_index_type bc_indices = createBoundaryIndexSpace( particles );
+    return BoundaryCondition<bc_index_type, UserFunctor>(
+        bc_indices, user_functor, force_update );
+}
+
+// Wrappers for single plane cases.
+template <class BoundaryType, class BCTag, class ExecSpace, class Particles>
+auto createBoundaryCondition( BCTag tag, const double value,
```
Comment (pabloseleson) [inline src/CabanaPD_Boundary.hpp:409]: Why do we need two "createBoundaryCondition" (Lines 409 and 419)?

File: examples/mechanics/crack_branching.cpp (hunk 1 above)
Comment (streeve) [reply examples/mechanics/crack_branching.cpp:99]: We would need a completely new option for that

File: examples/mechanics/crack_branching.cpp (hunk 2 above)
Comment (streeve) [reply examples/mechanics/crack_branching.cpp:104]: Unused time input

File: src/CabanaPD_Boundary.hpp (hunk 3 above)
Comment (streeve) [reply src/CabanaPD_Boundary.hpp:77]: Defaults, not hardcoded

File: src/CabanaPD_Boundary.hpp (hunk 5 above)
Comment (streeve) [reply src/CabanaPD_Boundary.hpp:409]: They have different inputs (tagged versions vs custom)

File: src/CabanaPD_Boundary.hpp (hunk 4 above)
Comment (streeve) [reply src/CabanaPD_Boundary.hpp:91]: Yes, that's possible but not something I'd thought to add. That would be possible with the new Custom version, but instead of passing that function here the user would just determine the particles on their own

File: examples/mechanics/crack_branching.cpp (hunk 2 above)
Comment (pabloseleson) [reply examples/mechanics/crack_branching.cpp:104]: Why do we need the unused time input?

File: examples/mechanics/kalthoff_winkler.cpp [hunk 6]
```diff
@@ -98,12 +98,11 @@ void kalthoffWinklerExample( const std::string filename )
     // ====================================================
     double dx = particles->dx[0];
     double x_bc = -0.5 * height;
-    CabanaPD::RegionBoundary plane(
+    CabanaPD::RegionBoundary<CabanaPD::RectangularPrism> plane(
         x_bc - dx, x_bc + dx * 1.25, y_prenotch1 - dx * 0.25,
```
Comment (pabloseleson) [inline examples/mechanics/kalthoff_winkler.cpp:102]: This would belong to a different PR, but why do we have 1.25?

File: src/CabanaPD_Boundary.hpp (hunk 3 above)
Comment (pabloseleson) [reply src/CabanaPD_Boundary.hpp:77]: So, if the user passes x_center and y_center as inputs, these are overwritten?

File: src/CabanaPD_Boundary.hpp [hunk 7]
```diff
@@ -304,12 +375,59 @@ auto createBoundaryCondition( UserFunctor user_functor, ExecSpace exec_space,
...
+                              const double initial_guess = 0.5 )
+{
+    std::vector<BoundaryType> plane_vec = { plane };
+    return createBoundaryCondition( tag, value, exec_space, particles,
+                                    plane_vec, initial_guess );
+}
+
+template <class UserFunctor, class BoundaryType, class ExecSpace,
+          class Particles>
+auto createBoundaryCondition( UserFunctor user_functor, ExecSpace exec_space,
```
Comment (pabloseleson) [inline src/CabanaPD_Boundary.hpp:421]: How do we use "user_functor"?

File: src/CabanaPD_Boundary.hpp [hunk 8]
```diff
@@ -159,15 +233,14 @@ struct BoundaryCondition
...
-                 RegionBoundary plane )
+    template <class ExecSpace, class Particles, class BoundaryType>
+    void update( ExecSpace exec_space, Particles particles, BoundaryType plane )
     {
         _index_space.update( exec_space, particles, plane );
     }
 
     template <class ExecSpace, class ParticleType>
-    void apply( ExecSpace, ParticleType&, double )
+    void apply( ExecSpace, ParticleType&, double time )
```
[the remark this answers is not in the document]
Comment (streeve) [reply src/CabanaPD_Boundary.hpp:262]: Any BC that is time dependent

File: src/CabanaPD_Boundary.hpp [hunk 9]
```diff
@@ -304,12 +375,59 @@ auto createBoundaryCondition( UserFunctor user_functor, ExecSpace exec_space,
...
+    return BoundaryCondition<bc_index_type, UserFunctor>(
+        bc_indices, user_functor, force_update );
+}
+
+// Wrappers for single plane cases.
+template <class BoundaryType, class BCTag, class ExecSpace, class Particles>
+auto createBoundaryCondition( BCTag tag, const double value,
+                              ExecSpace exec_space, Particles particles,
+                              BoundaryType plane,
+                              const double initial_guess = 0.5 )
```
[the remark this answers is not in the document]
Comment (streeve) [reply src/CabanaPD_Boundary.hpp:431]: You cannot know how many particles will be in the BC so you have to guess to allocate

File: src/CabanaPD_Boundary.hpp (hunk 7 above)
Comment (streeve) [reply src/CabanaPD_Boundary.hpp:440]: We use it in the crack branching

File: src/CabanaPD_Boundary.hpp (hunk 4 above)
Comment (streeve) [reply src/CabanaPD_Boundary.hpp:110]: Added in the most recent commit

File: src/CabanaPD_Boundary.hpp (hunk 3 above)
Comment (streeve) [reply src/CabanaPD_Boundary.hpp]: You could have before, but I expanded it a bit

**Review by pabloseleson — APPROVED** [review]
